core/src/test/java/org/archifacts/core/model/ApplicationBuilderTest.java

```diff
@@ -329,6 +335,43 @@ void assert_that_null_cannot_be_added_as_TargetBasedRelationshipDescriptor() {
 		}
 	}
 
+
+	@Nested
+	class ApplicationTests {
+
+		@Nested
+		@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+		class UnmodifiableCollections {
+
+			Arguments argument(String name, Function<Application, Set<?>> method) {
+				return Arguments.of(named(name, method));
+			}
+
+			Stream<Arguments> arguments() {
+				return Stream.of(
+					argument("getArtifacts", Application::getArtifacts),
+					argument("getBuildingBlocks", Application::getBuildingBlocks),
+					argument("getBuildingBlocksOfType", app -> app.getBuildingBlocksOfType(BuildingBlockType.of("TestType"))),
+					argument("getContainers", Application::getContainers),
+					argument("getContainersOfType", app -> app.getContainersOfType(ArtifactContainerType.of("TestType"))),
+					argument("getMiscArtifacts", Application::getMiscArtifacts),
+					argument("getExternalArtifacts", Application::getExternalArtifacts),
+					argument("getRelationships", Application::getRelationships),
+					argument("getRelationshipsOfRole", app -> app.getRelationshipsOfRole(ArtifactRelationshipRole.of("TestRole")))
+					);
+			}
+
+			@ParameterizedTest(name = "{0}")
+			@MethodSource("arguments")
+			void assert_that_collections_are_unmodifiable(Function<Application, Set<?>> setProvider) {
+				final Application application = Application
+				.builder()
+				.buildApplication(javaClasses);
+				Assertions.assertThatExceptionOfType(UnsupportedOperationException.class).isThrownBy(() -> setProvider.apply(application).add(null));
```

Comment:
While technically valid, I don't really like the adding of a _null_ value. I know that collections should throw NPEs if they don't allow _null_ values, but maybe we should consider using real objects here? If this inflates the test code too much, we can of course continue to use _null_.

Reply:
I agree that it's not optimal, but it ensures the test to be simple... otherwise I would need a `JavaClass` to build an artifact and things become much more complex without a real benefit.